wsdlgo/encoder.go

```diff
@@ -458,118 +455,152 @@ func (ge *goEncoder) writeGoFuncs(w io.Writer, d *wsdl.Definitions) error {
+{{- 	if ne .Name "err" }}
+	{{.Name}} = out.Body.{{fieldNameString .Name}}
+{{- 	end }}
+{{- end }}
+
+	return
 }
 `))
 
-func (ge *goEncoder) writeSOAPFunc(w io.Writer, d *wsdl.Definitions, op *wsdl.Operation, in, out, ret []string, xmlToken string) bool {
+func (ge *goEncoder) writeSOAPFunc(w io.Writer, d *wsdl.Definitions, op *wsdl.Operation, inParams, outParams []*parameter, messageName string) bool {
 	if _, exists := ge.soapOps[op.Name]; !exists {
 		return false
 	}
-	// TODO: handle other input params: e.g. SOAP headers.
-	if len(in) < 1 || len(out) != 2 {
-		return false
-	}
 	ge.needsStdPkg["encoding/xml"] = true
 	ge.needsExtPkg["github.com/fiorix/wsdl2go/soap"] = true
-	in[0] = renameParam(in[0], "α")
-	out[0] = renameParam(out[0], "β")
-	typ := strings.SplitN(out[0], " ", 2)
-	if strings.HasPrefix(ret[0], "&") {
-		ret[0] = "nil"
+
+	var soapAction string
+	soapOp := ge.soapOps[op.Name]
+	if soapOp.Operation != nil {
+		soapAction = soapOp.Operation.SoapAction
 	}
 	soapFuncT.Execute(w, &struct {
-		PortType      string
-		Name          string
-		Input         string
-		Output        string
-		OutputType    string
-		XMLOutputType string
-		RetPtr        bool
-		RetDef        string
+		PortType    string
+		Name        string
+		InParams    []*parameter
+		SoapAction  string
+		OutParams   []*parameter
+		MessageName string
 	}{
 		strings.ToLower(d.PortType.Name[:1]) + d.PortType.Name[1:],
 		strings.Title(op.Name),
-		strings.Join(in, ","),
-		strings.Join(out, ","),
-		strings.TrimPrefix(typ[1], "*"),
-		strings.TrimPrefix(xmlToken, "*"),
-		typ[1][0] == '*',
-		ret[0],
+		inParams,
+		soapAction, // TODO -- stick this into inParams above
+		outParams,
+		messageName,
 	})
 	return true
 }
 
-func renameParam(p, name string) string {
-	v := strings.SplitN(p, " ", 2)
-	if len(v) != 2 {
-		return p
-	}
-	return name + " " + v[1]
-}
-
 // returns list of function input parameters.
 func (ge *goEncoder) inputParams(op *wsdl.Operation) ([]*parameter, error) {
 	if op.Input == nil {
 		return []*parameter{}, nil
 	}
-	im := ge.trimns(op.Input.Message)
+	im := trimns(op.Input.Message)
 	req, ok := ge.messages[im]
 	if !ok {
 		return nil, fmt.Errorf("operation %q wants input message %q but it's not defined", op.Name, im)
 	}
-	return ge.genParams(req, true), nil
+
+	var parts []*wsdl.Part
+	if len(op.ParameterOrder) != 0 {
+		order := strings.Split(op.ParameterOrder, " ")
+
+		// Use a map to run O( len(parts)+len(order) )
+		partLookup := map[string]*wsdl.Part{}
+		for _, part := range req.Parts {
+			partLookup[part.Name] = part
+		}
+		for i, o := range order {
+			parts[i] = partLookup[o]
+		}
+
+	} else {
+		parts = req.Parts
+	}
+
+	return ge.genParams(parts, true), nil
 }
 
 // returns list of function output parameters plus error.
 func (ge *goEncoder) outputParams(op *wsdl.Operation) ([]*parameter, error) {
-	out := []*parameter{{code: "err error"}}
+	errP := &parameter{Name: "err", Type: "error", XMLName: "Err"}
```

Comment:
I had earlier test versions of this stick the error in the return struct--basically wasted space--this is a holdover from that.